File: rust/herb-linter/src/rules/actionview_no_silent_render.rs
use crate::utils::erb_utils::is_output_tag_opening;

use herb::nodes::ERBRenderNode;
use herb::Visitor;

rule_visitor!(ActionViewNoSilentRenderVisitor);
define_parser_rule!(
  ActionViewNoSilentRenderRule,
  "actionview-no-silent-render",
  Error,
  ActionViewNoSilentRenderVisitor,
  parser_options: { render_nodes: true }
);

impl Visitor for ActionViewNoSilentRenderVisitor {
  fn visit_erb_render_node(&mut self, node: &ERBRenderNode) {
    let tag_opening = node.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

    if !is_output_tag_opening(tag_opening) {
      self.add_offense(
        format!(
          "Avoid using `{} %>` with `render`. Use `<%= %>` to ensure the rendered content is output.",
          tag_opening
        ),
        node.location.clone(),
      );
    }

    self.walk_erb_render_node(node);
  }
}

File: javascript/packages/linter/src/rules/actionview-no-silent-render.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { isERBOutputNode } from "@herb-tools/core"

import type { ERBRenderNode, ParseResult, ParserOptions } from "@herb-tools/core"
import type { FullRuleConfig, LintContext, UnboundLintOffense } from "../types.js"

class ActionViewNoSilentRenderVisitor extends BaseRuleVisitor {
  visitERBRenderNode(node: ERBRenderNode): void {
    if (!isERBOutputNode(node)) {
      this.addOffense(
        `Avoid using \`${node.tag_opening?.value} %>\` with \`render\`. Use \`<%= %>\` to ensure the rendered content is output.`,
        node.location,
      )
    }

    this.visitChildNodes(node)
  }
}

export class ActionViewNoSilentRenderRule extends ParserRule {
  static ruleName = "actionview-no-silent-render"
  static introducedIn = this.version("0.9.1")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      render_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ActionViewNoSilentRenderVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
